aspnetcore/mvc/controllers/application-model.md

```diff
@@ -12,7 +12,187 @@ ms.technology: aspnet
 ms.prod: aspnet-core
 uid: mvc/controllers/application-model
 ---
-# 🔧 Working with the Application Model
+# Working with the Application Model
+
+By [Steve Smith](http://ardalis.com)
+
+ASP.NET Core MVC defines an *application model* representing the components of an MVC app. You can read and manipulate this model to modify how MVC elements behave.
+
+## Models and Providers
+
+The ASP.NET Core MVC application models include both abstract interfaces and concrete implementations that define how MVC behaves. This behavior includes how MVC determines and uses controller names, action names, action parameters, routes, and filters. By working with the application model, you can modify your app to follow different conventions from the default MVC behavior.
+
+The ASP.NET Core MVC Application Model has the following structure:
+
+* ApplicationModel
+	* Controllers (ControllerModel)
+		* Actions (ActionModel)
+			* Parameters (ParameterModel)
+
+Each level of the model has access to a common `Properties` collection, and lower levels can access and overwrite property values set by higher levels in the hierarchy.
```

Comment:
A few words about the properties collection....

The properties are persisted to the ActionDescriptor.Properties when the actions are created. Then when you're handling a request you can retrieve the data you attached in your convention through `ActionContext.ActionDescriptor.Properties`. 

Most of the per-request concepts in the framework get access to the AD. Using properties is a great way to configure your filters/modelbinders/whatever on a per-action basis.

We would often see in the past when users implement their own subsystem, they do it in such a way that don't have access to a configuration system as rich as the framework. Often developers end up doing a lot of reflection calls on a per-request basis because they have nowhere to stash the result. We absolutely intend for properties to be used as a cache for that kind of data. Just be aware that `ActionDescriptor.Properties` is not thread safe for writers once the startup is finished. Conventions are the best way to add data here.